content/articles/2017-11-15-withr-2.1.0.Rmd

```diff
@@ -0,0 +1,234 @@
+[graphics devices](http://withr.r-lib.org/reference/devices.html).
+
+These functions open a new graphics device, run some code, then automatically
+close the device. This is very useful when printing multiple plots to the same
+device.
+
+```{r}
+with_pdf("test.pdf", width = 7, height = 5, {
+  plot(mtcars$hp, mtcars$mpg)
+  plot(mtcars$wt, mtcars$mpg)
+})
+```
+
+![](test.pdf)
+
+Thanks to [Richard Cotton's](https://github.com/richierocks) great [pull
+request](https://github.com/r-lib/withr/pull/37) for this feature!
+
+## Connections
+
+There are two new functions for cleaning up connections in R.
+`with_connection()` allows you to automatically close R's file connections.
+Here we create a writable file connection, write some lines to it with
+`with_connection()`, then open a read-only connection and read the file using
+`local_connection()`.
+
+```{r}
+with_connection(list(con = file("temp", "w")), {
+  writeLines(c("foo", "bar"), con)
+})
+read_temp <- function() {
+  readLines(local_connection(file("temp", "r")))
+}
+read_temp()
+```
+```{r, echo = FALSE}
+unlink("temp")
+```
+
+`with_db_connection()` provides [DBI](http://rstats-db.github.io/DBI)
+connections to Relational Database Management Systems which automatically call
+`DBI::dbDisconnect()`. Here we create a new [SQLite](https://www.sqlite.org/)
+database, connect to it with `with_db_connection()`, and write a new table to
+it. We then create another connection with `local_db_connection()` and read
+from the table.
+
+```{r}
+db <- tempfile()
+with_db_connection(
+  list(con = DBI::dbConnect(RSQLite::SQLite(), db)), {
+    DBI::dbWriteTable(con, "mtcars", mtcars)
+})
+
+head_db_table <- function(...) {
+  con <- local_db_connection(DBI::dbConnect(RSQLite::SQLite(), db))
+  head(DBI::dbReadTable(con, "mtcars"), ...)
+}
+head_db_table()
+unlink(db)
+```
+
+## Packages / Namespaces / Environments
+
+Three new functions allow you to temporary attach a package, namespace or
+environment in a block of code.
+
+```{r}
+with_package("lattice", {
+  xyplot(y ~ x, data.frame(x = -2:2, y = dnorm(-2:2)))
+})
+```
+
+`with_namespace()` attaches the package namespace, so all objects (including
+unexported ones) are available. Here `vlapply` is an unexported functions from withr.
+
+```{r}
+with_namespace("withr", {
+  vlapply(1:5, function(x) x %% 2 == 0)
+})
+```
+
+`with_environment()` attaches any environment, detaching it in the same fashion
+as the other functions.
+
+
+## Other changes
+
+There are many other bug fixes and other minor improvements in this release.
+You can see a complete list in the [release
+notes](https://github.com/tidyverse/withr/releases/tag/v2.1.0).
+
+A big thanks goes to all the community members who contributed code and opened
+issues since the last release!
+
+```{r, eval = FALSE, include = FALSE}
+x <- gh::gh("/repos/:owner/:repo/issues", owner = "r-lib", repo = "withr", since = "2017-06-16", state = "all")
```

Comment:
Does this include PRs?

Reply:
Yes it should (https://developer.github.com/v3/issues/#list-issues)

>  As far as the API is concerned, every pull request is an issue, but not every issue is a pull request.

Reply:
Awesome - this is way simpler than my code! @batpigandme